docs/ps_modules/latest_news.py

```diff
@@ -0,0 +1,176 @@
+@dataclass(frozen=True)
+class Article:
+    filename: str
+    title: str
+    date: datetime
+    keywords: List[str]
+    description: str
+    priority: int
+
+    @property
+    def datef(self) -> str:
+        return self.date.strftime("%d %b %Y")
+
+    def summary(self, width: int = 140) -> str:
+        return shorten(self.description, width=width, placeholder="...")
+
+    @property
+    def link(self) -> str:
+        return f"latest_news/{self.filename}"
+
+
+# ------------------------------ Parsing ---------------------------------------
+
+def _first_nonempty_paragraph(doc: nodes.document) -> Optional[str]:
+    p = doc.next_node(nodes.paragraph)
+    return p.astext().strip() if p else None
+
+def _meta(doc: nodes.document) -> Dict[str, str]:
+    out: Dict[str, str] = {}
+    for m in doc.findall(nodes.meta):
+        name = (m.get("name") or "").strip().lower()
+        val = (m.get("content") or "").strip()
+        if name and val:
+            out[name] = val
+    return out
+
+def parse_article(path: Path) -> Article:
+    doc = publish_doctree(path.read_text(encoding="utf-8"))
+    meta = _meta(doc)
+
+    title = (meta.get("title") or path.stem).strip()
+
+    priority = meta.get("priority")
+    if priority and priority.strip():
+        priority = int(priority)
+    else:
+        priority = 100
+
+    date_raw = (meta.get("date") or "").strip()
+    if not date_raw:
+        raise ValueError(f"{path}: missing required 'date' meta")
+    date = datetime.strptime(date_raw, "%d-%m-%Y")
+
+    kw = [t.strip() for t in (meta.get("keywords", "")).split(",") if t.strip()]
+    desc = _first_nonempty_paragraph(doc) or ""
+
+    return Article(
+        filename=path.stem, title=title, date=date, keywords=kw, description=desc, priority=priority
+    )
+
+
+# ------------------------------ Rendering -------------------------------------
+IND_TAB = " " * 3
+IND_CARD = " " * 6
+
+EMPTY_NOTE = dedent("""\
+.. note::
+   No news posts available.
+
+""")
+
+CARD_TMPL = dedent("""\
+.. card::
+   :link: {link}
+   :link-type: doc
+   :shadow: none
+
+   :fas:`calendar-alt` {datef}
+
+   **{title}**
+
+   {description}
+""").rstrip()
+
+TABSET_HEADER = dedent("""\
+.. tab-set::
+""").rstrip()
+
+FIRST_TAB = dedent("""\
+.. tab-item:: All
+   :selected:
+""").rstrip()
+
+TAB_ITEM_TMPL = ".. tab-item:: {keyword}"
+
+CAROUSEL_HEADER = dedent("""\
+.. card-carousel:: 3
+""").rstrip()
+
+
+
+def render_carousel(posts: Iterable[Article]) -> str:
+    posts = list(posts)
+    if not posts:
+        return EMPTY_NOTE
+
+    parts = [CAROUSEL_HEADER, ""]
+    for a in posts:
+        card = CARD_TMPL.format(
+            link=a.link, datef=a.datef, title=a.title, description=a.description
+        )
+        parts.append(indent(card, IND_TAB))
+
+    return "\n\n".join(parts).rstrip() + "\n"
+
+def render_tabset(dir_path: Path) -> str:
+    all_posts, all_keywords = build_index(dir_path, keyword_filter=None)
+
+    title = dedent("""\
```

Comment:
now you use dedent everywhere, but I think it does nothing, it removes all common whitespace up to the first character. You never have that, so in principle it doesn't do anything!